### ISSUE TITLE
How do you get frequencies from HarmonicAnalyser harmonics?

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#302]
```cpp
HarmonicAnalyser analyzer;
vector<int> harmonics = analyzer.getIndicesOfMajorHarmonics(doubleVec, analysisPointInSecs * sampleRate, blockSize);
```

What are the frequencies for said harmonics?

**Comment by RobinSchmidt** [comment]
you want the major harmonic frequencies of a given block? ...and "major" in the sense of "loudest" (not as in major or minor chords)? the rsHarmonicAnalyzer is actually supposed to be used offline on a whole signal - not block-wise. it delivers an rsSinusoidalModel object. if you want to extract the most important harmonics/partials from that, i could perhaps implement a function that computes the average or rms amplitude for each partial (averaged over time) - like a total energy of each partial - and returns all the total energies. then we could sort that array (along with an index array) and extract the most energetic partials. ...if that's what you want

or wait...no - it seems, you want the indices of the most important partials at a given time-instant? ...and you want to extract them from a "doubleVec", i.e. a raw signal?

**Comment by elanhickler** [comment]
This is for noise reductgion.

The end goal is to have a signal of harmonics ONLY. This is notoriously inconsistent and messy no matter what algorithm. So, what we really have to do is get major harmonics and guess where the rest of them are (the ones that fall below whatever threshold).

So let's say we have a fundamental at 100hz and the algorithm detects 100, 200, 300, 400, 600. Oops, we missed 500. So we need to assume there's a harmonic at 500. Also, above 600, we assume there is one at 700, 800, 900, 1000, all the way up to sampleRate / 2.

Now, the problem is that signals we work with are not so simple. In reality, the harmonics will be something like 100, 198, 296, 394, as in, they are not perfect multiples and "drift off the perfect multiple" the higher in frequency you go.

In other words, we need a whole algorithm solution to make a good guess as to where all the harmonics are at one instance of time.

After that, the goal is to guess where the harmonics end in time. This is not crucial because I can just take a lowpass filter and lower the cutoff as the signal continues to remove noise where harmonics are no longer ringing.

**Comment by RobinSchmidt** [comment]
hmm....actually, the algorithm already searches for the partial near the expected harmonic frequencies. are you talking about the (old) problem that sometimes the algo misses a partial (i.e sets its amplitude to zero when it should be nonzero?). it seems to me that we should probably refine the criteria for what is and what isn't a partial

**Comment by elanhickler** [comment]
For now, could you tell me how to get the frequency values? That's the first step.

**Comment by RobinSchmidt** [comment]
i'm not sure, if i really understand what you want, but if you want the frequency trajectory of a given partial, you would create a sinusoidal model object as usual, like so

`RAPT::rsSinusoidalModel<double> mdl = analyzer.analyze(&x[0], (int) x.size());`

and then you could do something like

`std::vector<double> freqs = mdl.getPartial(k).getFrequencyArray();`

this would give you an array of the measured frequencies of partial k. but i have no idea what you are going to do with that. however, the frequencies are probably useless, without having the time-stamps for the datapoints also, which you can get via

`std::vector<double> timeStamps = mdl.getPartial(k).getTimeArray();`

(i did not test this code - this is just, how its supposed to work)

**Comment by elanhickler** [comment]
I tried the sinusoidal analyzer and it gives hundreds of thousands of partials. I'm looking for a number of partials equal to 20,000 (limit of human hearing) divided by fundamental frequncy. For example if the fundamental frequency is 100, then I will want 200 partials or less, whatever is above the noise floor.

I tried `removePartialsWithMeanFreqAbove` and it takes forever. maybe a half an hour for one rhodes note. That reduced the number of partials to around 150,000.

Most of the partials' frequency arrays were a size of 3, so it only had 3 frequency data points.

I need something more geared towards getting the entirely of every harmonic detected in a signal.

**Comment by RobinSchmidt** [comment]
oh - that definitely sounds like a bug. oh....wait are you talking about class rsSinusoidalAnalyzer (and not rsHarmonicAnalyzer)? because the former is still in somewhat early stages of construction.

**Comment by elanhickler** [comment]
what can rsHarmonicAnalyzer output in terms of tracking harmonics?

**Comment by elanhickler** [comment]
I just tried HarmonicAnalyzer.analyze and it crash at rsArrayTools::copy because destination is nullptr

**Comment by elanhickler** [comment]
trying to update my rs-met, but get compile errors

cannot open source file "BasicFunctions.h"
cannot open source file "Trees.cpp"
cannot open source file "Trees.h"
cannot open source file "Scientific/QuantileFilter.cpp"
cannot open source file "Scientific/QuantileFilter.h"
cannot open source file "Functions/BasicMathFunctions.cpp"
cannot open source file "Types/Fraction.h"
cannot open source file "Functions/BasicMathFunctions.h"

**Comment by elanhickler** [comment]
Trying to separate harmonics. I can do it easily by hand in RX, but when I zero out bins using your FFT functions I get weird artifacts.

![iZotope_RX_3_(64-bit)_2020-08-29_18-36-10](https://user-images.githubusercontent.com/9116575/91649230-981e4980-ea26-11ea-8ab0-760155132a69.png)
![2020-08-29_18-36-44](https://user-images.githubusercontent.com/9116575/91649231-9a80a380-ea26-11ea-9a29-6b9efbfe523e.png)

```cpp
template<typename T> class ElanFFT
{
public:
	static RAPT::rsMatrix<std::complex<T>> getDecayingHarmonicProfile(const vector<T>& signal, int sampleRate, double startTime, int blockSize, double decibelThreshold = -80)
	{
		RAPT::rsMatrix<std::complex<T>> mat;
		RAPT::rsSpectrogramProcessor<T> sp;
		sp.setBlockAndTrafoSize(blockSize, blockSize * 2);
		sp.setHopSize(blockSize / 4);
		sp.setAnalysisWindowType(WindowType::blackman);
		sp.setSynthesisWindowType(WindowType::blackman);
		sp.setOutputDemodulation(true);

		mat = sp.getComplexSpectrogram(signal.data(), signal.size());
		
		vector<vector<int>> harmonicProfile;

		double threshold = pow(10.0, decibelThreshold / 20.0);

		for (int c = 0; c < mat.getNumRows(); ++c)
		{
			std::vector<int> harmonics;
			for (int r = 4; r < (int(mat.getNumColumns()) / 2) - 4; ++r)
			{
				bool condition =
					abs(mat(c, r)) > abs(mat(c, r - 4)) &&
					abs(mat(c, r)) > abs(mat(c, r - 3)) &&
					abs(mat(c, r)) > abs(mat(c, r - 2)) &&
					abs(mat(c, r)) > abs(mat(c, r - 1)) &&
					abs(mat(c, r)) > abs(mat(c, r + 1)) &&
					abs(mat(c, r)) > abs(mat(c, r + 2)) &&
					abs(mat(c, r)) > abs(mat(c, r + 3)) &&
					abs(mat(c, r)) > abs(mat(c, r + 4)) &&
					abs(mat(c, r)) > threshold;

				if (condition)
					harmonics.push_back(r);
			}

			harmonicProfile.push_back(harmonics);
		}

		removeErroneousHarmonics(harmonicProfile, 3);
		removeHarmonicDropouts(harmonicProfile);
		padHarmonicEnds(harmonicProfile, sampleRate, blockSize, 0.5 * 4);

		for (auto& c : harmonicProfile)
			std::sort(c.begin(), c.end(), [](int a, int b) { return a < b; });

		//vector<int> featherNums = { 5, 4, 3, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5};
		vector<int> featherNums = { 3, 2, 1, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3};
		int divideBy = featherNums.back() + 1;

		for (int c = 0; c < harmonicProfile.size(); ++c)
		{
			for (int r = 0; r < harmonicProfile[c].size(); ++r)
			{
				int f = -featherNums.size() / 2;
				for (int i = 0; i < featherNums.size(); ++i)
				{
					int row = harmonicProfile[c][r] + f;
					if (juce::isPositiveAndBelow(row, mat.getNumColumns()))
						mat(c, row) *= T(featherNums[i]) / T(divideBy);
					++f;
				}
			}
		}

		return std::move(mat);
	}

private:

	static int getNumBlocksForSignal(vector<T> signal, int blockSize)
	{
		return signal.size() / blockSize;
	}

	static int getStartingBlockFromTime(double startTime, int sampleRate, int blockSize)
	{
		return (startTime * sampleRate) / blockSize;
	}

	static void removeErroneousHarmonics(vector<vector<int>>& harmonicProfile, int maxDropouts)
	{
		vector<int> allHarmonicsFound;

		for (int c = 0; c < harmonicProfile.size(); c++)
			for (int r = 0; r < harmonicProfile[c].size(); ++r)
				VectorHelper::pushUnique(allHarmonicsFound, harmonicProfile[c][r]);

		for (auto& h : allHarmonicsFound)
		{
			int dropoutColumn = 0;
			int numberOfDropouts = 0;
			for (int c = 0; c < harmonicProfile.size(); c++)
			{
				if (!VectorHelper::contains(harmonicProfile[c], h))
					++numberOfDropouts;

				if (numberOfDropouts > maxDropouts)
				{
					dropoutColumn = c;
					break;
				}				
			}

			if (numberOfDropouts > maxDropouts)
				for (int c = dropoutColumn; c < harmonicProfile.size(); c++)
					VectorHelper::removeElement(harmonicProfile[c], h);
		}
	}

	static void removeHarmonicDropouts(vector<vector<int>>& harmonicProfile)
	{
		vector<int> harmonicsToAdd;
		for (int c = harmonicProfile.size(); c--; )
		{
			for (int r = 0; r < harmonicProfile[c].size(); ++r)
				VectorHelper::pushUnique(harmonicsToAdd, harmonicProfile[c][r]);

			for (auto& h : harmonicsToAdd)
				VectorHelper::pushUnique(harmonicProfile[c], h);
		}
	}

	static void padHarmonicEnds(vector<vector<int>>& harmonicProfile, int sampleRate, int blocksize, double seconds)
	{
		int paddingInBlocks = (sampleRate * seconds) / blocksize;

		paddingInBlocks = std::min<int>(paddingInBlocks, harmonicProfile.size());

		vector<int> harmonicsAlreadyPadded;

		for (int c = harmonicProfile.size(); c--; )
		{
			for (int r = 0; r < harmonicProfile[c].size(); ++r)
			{
				int h = harmonicProfile[c][r];

				if (VectorHelper::contains(harmonicsAlreadyPadded, h))
					continue;

				for (int c2 = c + 1; c2 < harmonicProfile.size() && c2 < c + 1 + paddingInBlocks; ++c2)
					harmonicProfile[c2].push_back(h);

				VectorHelper::pushUnique(harmonicsAlreadyPadded, h);
			}
		}
	}
};
```

**Comment by elanhickler** [comment]
I have hired a programmer to create a FIR filter to do harmonic extraction. Looks like zeroing out bins is not the correct way to do this, so it's not an issue with your FFT library. Closing this thread.
